packages/twenty-server/src/engine/core-modules/auth/auth.resolver.ts

```diff
@@ -111,6 +111,13 @@ export class AuthResolver {
       );
     }
     const user = await this.authService.challenge(challengeInput, workspace);
+
+    if (user.totpSecret)
```

Comment:
Maybe you could instead ask for the totp in the method that enables exchanging a loginToken against an auth token pair (access + refresh)? That way even someone who logins through microsoft could add an additional 2FA layer if they want to.  This challenge method is only for email/pwd login

Reply:
I think that in case of SSO, 2FA is handled by the identity provider. I don't feel that it's a usual behaviour to have an additional "manual" 2FA.
IMO this 2FA should be for the email/password identity provider only.

For instance with Google Auth, authentication method can include TOTP, Passkeys, Hardware devices, etc.

I can make further research of SSO with additional 2FA if you think it's a good pattern.

Reply:
I get your point but that's not my experience everywhere. For example on Airbnb I have a 2FA after Google login.  Same on coinbase, after Google login it's asking me for a second factor. 

Also if you set 2FA at the login level then it's harder to put 2FA at the userWorkspace level (because you could have login with email/pwd -> redirect to the right workspace -> 2FA)

I agree that if it's a company SSO then it usually doesn't make sense to configure an additional 2FA. But I don't see any downside to putting it as a separate step? It feels like it would be clean and can only bring additional capabilities. Am I missing a downside?

Reply:
Yes great questions all the way, didn't expect there would be so much to uncover, it's always the same haha :))